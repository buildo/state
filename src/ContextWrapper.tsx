import * as React from 'react';
import * as PropTypes from 'prop-types';

export type ProvideContext = { [k: string]: any };
export type ProvideContextTypes = PropTypes.ValidationMap<any>;

export default function mkContextWrapper(ctx: ProvideContext, ctxTypes: ProvideContextTypes): React.ComponentType<{ children: () => JSX.Element }> {
  return class ContextWrapper extends React.Component<{ children: () => JSX.Element }> {
    static childContextTypes = ctxTypes

    getChildContext() {
      return ctx;
    }

    render() {
      return this.props.children();
    }
  }
}
