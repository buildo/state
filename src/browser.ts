import createHistory from 'history/createBrowserHistory';
import trim = require('lodash/trim');
import omit = require('lodash/omit');
import * as qs from 'qs';

export type BrowserState = { [k: string]: string } & { view: string };

// TODO: makes it hardly testable and also not configurable for iso (needs to be static location + not a singleton)
const history = createHistory();

// TODO: below, ignoring path params for the moment
// there's only `view` which is the whole `pathname` from history perspective

export function syncToBrowser(s: BrowserState, push: boolean = true): void {
  const pathname = `/${s.view}`;
  const search = `?${qs.stringify(omit(s, ['view']))}`;
  (push ? history.push : history.replace)({ pathname, search });
}

function parseBrowserState(location: typeof history.location): BrowserState {
  const view = trim(location.pathname, ' /');
  const s = qs.parse(location.search);
  return { ...s, view };
}

export function onBrowserChange(callback: (s: BrowserState) => void): void {
  history.listen(location => {
    setTimeout(() => {
      // TODO: something better than setTimeout here?
      callback(parseBrowserState(location));
    });
  });
  setTimeout(() => {
    // TODO: something better than setTimeout here?
    callback(parseBrowserState(history.location));
  });
}
