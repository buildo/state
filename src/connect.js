import React from 'react';
import debug from 'debug';
import { filter } from 'rxjs/operator/filter';
import { map } from 'rxjs/operator/map';
import omit from 'lodash/omit';
import identity from 'lodash/identity';
import pick from 'lodash/pick';
import t from 'tcomb';
import shallowEqual from './shallowEqual';

const warn = debug('state:connect');
warn.log = ::console.warn; // eslint-disable-line no-console

export const ConnectContextTypes = {
  transition: React.PropTypes.func.isRequired,
  state: React.PropTypes.object.isRequired
};

const defaultGetNewState = ty => v => ty.reduce((acc, key) => ({
  ...acc,
  [key]: v[key]
}), {});

// expects a select function or a list of keys
// and an optional configuration object
export default stateType => (select = identity, {
  // implement a standard shouldComponentUpdate with shallowEquals
  // do not use on non-pure components, e.g. react-router `RouteHandler`s
  //
  // Boolean
  //
  pure = true,

  // do not update the component props if the type declaration doesn't match
  // useful to skip "router transitioning" problems: old container that should be unmounted
  // soon (but still async) by re-rendering after a router change would otherwise receive
  // the new (and possibly incorrect from their POV) state
  // TODO(gio): elaborate better / try to remove
  //
  // Boolean
  //
  filterValid = true,

  // optionally kill some props using lodash omit
  // useful to stay strict in @props declaration, a typical default could be:
  // ['router', 'query', 'params']
  //
  // Array<String>
  //
  killProps = []
} = {}) => {

  const isKeyList = t.list(t.String).is(select);
  if (!isKeyList && !t.Function.is(select)) {
    throw new Error('connect expects a select function or a list of keys');
  }

  select.forEach(k => {
    if (!stateType.meta.props.hasOwnProperty(k)) {
      throw new Error(`${k} is not defined in state`);
    }
  });

  const decorator = Component => {
    const displayName = `connect(${Component.displayName || Component.name || 'Component'})`;

    const getNewState = isKeyList ? defaultGetNewState(select) : select;

    const shouldUpdateState = filterValid && isKeyList ? v => {
      const invalid = select.reduce((acc, k) => {
        return !stateType.meta.props[k].is(v[k]) ? [...acc, k] : acc;
      }, []);

      if (invalid.length > 0 && process.env.NODE_ENV === 'development') {
        warn(`Skipping update for ${displayName}. Invalid keys: ${invalid.join(', ')}`);
      }
      return invalid.length === 0;
    } : () => true;

    return class ConnectWrapper extends React.Component {
      static contextTypes = ConnectContextTypes

      static displayName = displayName;

      constructor(props, context) {
        super(props, context);
        const value = context.state.value;
        if (shouldUpdateState(value)) {
          this.state = getNewState(value);
        } else {
          this.state = {};
        }
      }

      componentDidMount() {
        this._subscription = this.context.state::filter(shouldUpdateState)::map(getNewState).subscribe(::this.setState);
      }

      componentWillUnmount() {
        if (this._subscription) {
          this._subscription.unsubscribe();
        }
      }

      shouldComponentUpdate(nextProps, nextState) {
        return pure ? !shallowEqual(this.props, nextProps) || !shallowEqual(this.state, nextState) : true;
      }

      render() {
        const props = {
          // state first: connected props are overridable by "actual" (passed) props
          // it's useful for av@queries, av@commands
          ...this.state,
          ...omit(this.props, killProps),
          transition: this.context.transition
        };
        return (
          <Component {...props} />
        );
      }
    };
  };
  if (isKeyList) {
    decorator.Type = { ...pick(stateType.meta.props, select), transition: t.Function };
  }
  return decorator;
};
