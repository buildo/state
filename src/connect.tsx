import * as React from 'react';
import * as PropTypes from 'prop-types';
import * as debug from 'debug';
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/map';
import { BehaviorSubject } from 'rxjs/BehaviorSubject';
import { Subscription } from 'rxjs/Subscription';
import omit = require('lodash/omit');
import pick = require('lodash/pick');
import * as t from 'tcomb';
import shallowEqual from './shallowEqual';
import { TransitionFunction } from './transition';
import { ObjectOmit } from 'typelevel-ts';
import { StateT as ST } from './StateT';

declare var process: any; // TODO(typo)

const warn = debug('state:connect');
warn.log = console.warn.bind(console); // eslint-disable-line no-console

export const ConnectContextTypes = {
  transition: PropTypes.func.isRequired,
  state: PropTypes.object.isRequired
};

function defaultGetNewState<S extends ST, Decl extends string>(declaration: Decl[]) {
  return (v: S) =>
    declaration.reduce(
      (acc, key) => ({
        ...acc,
        [key as string]: v[key as string] // TODO(typo)
      }),
      {}
    );
}

export type ConnectConfig = {
  /**
   * Implement a standard shouldComponentUpdate with shallowEquals.
   * Should be `false` on non-pure components, e.g. react-router `RouteHandler`s.
   * default: true
   */
  pure?: boolean;
  /**
   * Do not rerender if received state is not valid.
   * default: true
   * @deprecated not needed anymore (it was needed with React Router)
   */
  filterValid?: boolean;
  /**
   * Do not pass down some of the received props.
   * default: []
   * @deprecated not needed anymore (it was handy with React Router)
   */
  killProps?: string[];
};
export type ConnectedComponentProps<P extends {}, Decl extends string> = ObjectOmit<P, Decl | 'transition'>;
export type ConnectedComponent<P extends {}, Decl extends string> = React.ComponentType<
  ConnectedComponentProps<P, Decl>
>;
export type DecoratedComponent<S extends ST, P extends {}> = React.ComponentType<
  P & { transition?: TransitionFunction<S> }
>;
export type ConnectDecorator<S extends ST, Decl extends keyof S> = <P>(
  /**
   * Component to connect.
   * Will be rendered adding the declared keys from state and `transition` as props
   */
  C: DecoratedComponent<S, P>
) => ConnectedComponent<P, Decl>;
export type ConnectDeclaration<S extends ST, Decl extends keyof S> = ConnectDecorator<S, Decl> &
  ({
    Type: { [k in Decl]: t.Type<any> } & { transition: t.Function1 };
  });
export type Connect<S extends ST> = <Decl extends keyof S>(
  /**
   * Keys of State to be connected and passed to this container
   */
  decl: Decl[],
  /**
   * Additional configurations
   */
  config?: ConnectConfig
) => ConnectDeclaration<S, Decl>;

// global state tcomb type
export default function makeConnect<S extends ST>(stateType: t.Interface<S>): Connect<S> {
  // expects a list of keys to connect from state
  // and an optional configuration object
  return function connectDeclaration<Decl extends keyof S>(
    decl: Decl[] = [],
    {
      pure = true,
      // do not update the component props if the type declaration doesn't match
      // useful to skip "router transitioning" problems: old container that should be unmounted
      // soon (but still async) by re-rendering after a router change would otherwise receive
      // the new (and possibly incorrect from their POV) state
      // TODO(gio): elaborate better / try to remove
      filterValid = true,
      killProps = []
    }: ConnectConfig = {}
  ) {
    if (process.env.NODE_ENV !== 'production') {
      const isKeyList = t.list(t.String).is(decl);
      if (!isKeyList) {
        throw new Error('connect expects a list of state keys');
      }

      decl.forEach(k => {
        if (!stateType.meta.props.hasOwnProperty(k)) {
          throw new Error(`${k} is not defined in state`);
        }
      });
    }

    const decorator: ConnectDecorator<S, Decl> = <P extends {}>(Component: DecoratedComponent<S, P>) => {
      const displayName = `connect(${Component.displayName || Component.name || 'Component'})`;

      const getNewState = defaultGetNewState(decl);

      const shouldUpdateState = filterValid
        ? (v: S) => {
            const invalid = decl.filter(k => !(stateType.meta.props[k] as t.Type<any>).is(v[k])); // TODO(typo)

            if (invalid.length > 0 && process.env.NODE_ENV === 'development') {
              warn(`Skipping update for ${displayName}. Invalid keys: ${invalid.join(', ')}`);
            }
            return invalid.length === 0;
          }
        : () => true;

      type ContextT<S extends ST> = {
        // TODO(typo): move elsewhere and reuse
        state: BehaviorSubject<S>;
        transition: TransitionFunction<S>;
      };
      type StateT<S extends ST, Decl extends keyof S> = { [k in Decl]: S[k] } | {};
      return class ConnectWrapper extends React.Component<ConnectedComponentProps<P, Decl>, StateT<S, Decl>> {
        static contextTypes = ConnectContextTypes;

        static displayName = displayName;

        constructor(props: ConnectedComponentProps<P, Decl>, context: ContextT<S>) {
          super(props, context);
          const value = context.state.value;
          if (shouldUpdateState(value)) {
            this.state = getNewState(value);
          } else {
            this.state = {};
          }
        }

        context: ContextT<S>;
        _subscription?: Subscription;

        componentDidMount() {
          this._subscription = this.context.state
            .filter(shouldUpdateState)
            .map(getNewState)
            .subscribe(this.setState.bind(this));
        }

        componentWillUnmount() {
          if (this._subscription) {
            this._subscription.unsubscribe();
          }
        }

        shouldComponentUpdate(nextProps: ConnectedComponentProps<P, Decl>, nextState: StateT<S, Decl>) {
          return pure ? !shallowEqual(this.props, nextProps) || !shallowEqual(this.state, nextState) : true;
        }

        render() {
          const props = {
            // state first: connected props are overridable by "actual" (passed) props
            // it's useful for av@queries, av@commands
            ...this.state as any, // TODO(typo): spread types can only be created from object types
            ...omit(this.props as any, killProps), // TODO(typo): spread types can only be created from object types
            transition: this.context.transition
          };
          return <Component {...props} />;
        }
      };
    };

    (decorator as ConnectDeclaration<S, Decl>).Type = {
      ...pick(stateType.meta.props, decl) as any, // TODO(typo): ugh
      transition: t.Function
    };
    return decorator as ConnectDeclaration<S, Decl>;
  };
}
