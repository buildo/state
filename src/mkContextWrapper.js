import React from 'react';

export default function mkContextWrapper(ctx, ctxTypes) {
  return class ContextWrapper extends React.Component {
    static childContextTypes = ctxTypes

    getChildContext() {
      return ctx;
    }

    render() {
      return this.props.children();
    }
  };
}
