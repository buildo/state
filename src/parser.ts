import * as t from 'tcomb';
import { StateT, StateTcombType } from './StateT';
import { BrowserState } from './browser';
import mapValues = require('lodash/mapValues');
import identity = require('lodash/identity');

type TypeName = 'boolean' | 'number' | 'default';

function getIrreducibleTypeName(type: t.Irreducible<any>): TypeName {
  switch (type.meta.name) {
    case 'Number':
      return 'number';
    case 'Boolean':
      return 'boolean';
    default:
      return 'default';
  }
}

function getTypeName(type: t.Type<any>): TypeName {
  if (type.meta.kind === 'irreducible') {
    return getIrreducibleTypeName(type as t.Irreducible<any>);
  } else if (type.meta.kind === 'maybe') {
    const mType = (type as t.Maybe<any>).meta.type as t.Type<any>;
    return mType.meta.kind === 'irreducible' ? getIrreducibleTypeName(mType as t.Irreducible<any>) : 'default';
  }
  return 'default';
}

function parseBoolean(value: string): boolean | null {
  if (value === 'true') {
    return true;
  }
  if (value === 'false') {
    return false;
  }
  return null;
}

function parseNumber(value: string): number | null {
  const parsedValue = parseFloat(value);
  if (String(parsedValue) === value) {
    return parsedValue;
  }
  return null;
}

const parsers: { [k in TypeName]: (v: string) => any } = {
  boolean: parseBoolean,
  number: parseNumber,
  default: identity
};

export type ParsedBrowserState = BrowserState & { [k: string]: any };

export function stringify<S extends StateT>(_stateType: StateTcombType<S>) {
  return (params: ParsedBrowserState): BrowserState => {
    return mapValues<any, string>(params, String) as BrowserState;
  };
}

export function parse<S extends StateT>(stateType: StateTcombType<S>) {
  return (params: BrowserState): ParsedBrowserState => {
    return Object.keys(params).reduce((acc, paramName) => {
      const paramType = stateType.meta.props[paramName] as t.Type<any>;
      if (!paramType) {
        return acc;
      }
      const paramTypeName = getTypeName(paramType);
      const param = params[paramName];
      const parse = parsers[paramTypeName];
      return {
        ...acc,
        [paramName]: parse(param)
      };
    }, {}) as ParsedBrowserState;
  };
}
