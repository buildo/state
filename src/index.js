import run from './run';
import connect from './connect';
import t from 'tcomb';

const isStrictInterface = (x) => {
  return t.isType(x) && (x.meta.kind === 'interface') && (x.meta.strict === true);
};

export default stateType => {
  if (!isStrictInterface(stateType)) {
    throw new Error('`stateType` must be a strict tcomb interface');
  }

  return {
    run: run(stateType),
    connect: connect(stateType)
  };
};
