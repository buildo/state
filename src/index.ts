import * as t from 'tcomb';
import run, { Run } from './run';
import connect, { Connect } from './connect';
import { StateT } from './StateT';

const isStrictInterface = (x: any) => {
  return t.isType(x) && x.meta.kind === 'interface' && x.meta.strict === true;
};

export default function<S extends StateT>(
  _stateType: t.Interface<S>
): {
  run: Run<S>;
  appState: t.Interface<S>;
  connect: Connect<S>;
} {
  if (!isStrictInterface(_stateType)) {
    throw new Error('`stateType` must be a strict tcomb interface');
  }

  const stateType = t.interface(
    {
      ___k: t.maybe(t.Number),
      ..._stateType.meta.props
    },
    { strict: true, name: _stateType.meta.name }
  ) as t.Interface<S>;

  return {
    appState: stateType,
    run: args => run(stateType)(args),
    connect: connect(stateType)
  };
}
