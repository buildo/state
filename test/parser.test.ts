import { parse as _parse, stringify as _stringify } from '../src/parser';
import * as t from 'tcomb';

type S = {
  view: string;
  numb: number;
  bool: boolean;
  optNumb?: number;
  optBool?: boolean;
};
const ST = t.interface<S>({
  view: t.String,
  numb: t.Number,
  bool: t.Boolean,
  optNumb: t.maybe(t.Number),
  optBool: t.maybe(t.Boolean)
});
const parse = _parse(ST);
const stringify = _stringify(ST);

describe('parse', () => {
  it('should parse numbers', () => {
    expect(parse({ view: 'foo', numb: '4' })).toMatchSnapshot();
  });
  it('should parse optional numbers', () => {
    expect(parse({ view: 'foo', optNumb: '4' })).toMatchSnapshot();
  });
  it('should parse booleans', () => {
    expect(parse({ view: 'foo', bool: 'true' })).toMatchSnapshot();
  });
  it('should parse optional booleans', () => {
    expect(parse({ view: 'foo', optBool: 'false' })).toMatchSnapshot();
  });
});
