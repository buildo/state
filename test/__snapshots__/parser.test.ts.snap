// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parse should parse booleans 1`] = `
Object {
  "bool": true,
  "view": "foo",
}
`;

exports[`parse should parse numbers 1`] = `
Object {
  "numb": 4,
  "view": "foo",
}
`;

exports[`parse should parse optional booleans 1`] = `
Object {
  "optBool": false,
  "view": "foo",
}
`;

exports[`parse should parse optional numbers 1`] = `
Object {
  "optNumb": 4,
  "view": "foo",
}
`;
